kong/core/handler.lua

```diff
@@ -117,14 +119,12 @@ return {
           "' with: "..tostring(err))
       end
 
-      if balancer_address.hostname and not api.preserve_host then
-        var.upstream_host = balancer_address.hostname
+      -- if set `host_header` is the original header to be preserved
+      var.upstream_host = host_header or 
```

Comment:
I more and more find this syntax to be harder to read. In the future we should stick to:
```lua
if host_header then
  var.upstream_host = host_header

else
  var.upstream_host = balancer_address.hostname..":"..balancer_address.port
end
```